### PR TITLE
base codes for enumerables

### PR DESCRIPTION
**Pull request by danielochuba**
# Add own Enumerables 
### In this project we implemented enumerables by building them through the following processes:
- Created a class MyList and a module MyEnumerable
- Used a module inside the  `MyList`  class. 
- The `MyEnumerable` module will implement a subset of the functionality of [Enumerable](https://ruby-doc.org/core-3.0.0/Enumerable.html).

### CODE REVIEW

File: my_list.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+require './my_enumerable'
```
Comment (bobb-Rob) [inline my_list.rb:1]: - [ ] You are expected to use `require_relative` to import `MyEnumerable`. Using `require_relative` is usually for files in your directory, instead of `require`  as this is a ruby convention. "Require should be used for external files, like gems, while `require_relative` should be used for referring to files within your directory." Please fix this

File: my_list.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+class MyList
+  include MyEnumerable
+  def initialize(*args)
+    @list = args
+  end
+
+  def each(&block)
+    @list.each(&block)
+  end
```
Comment (bobb-Rob) [inline my_list.rb:11]: **[OPTIONAL]** Great job on implementing the #each method for the MyList class! Your current approach seems to work well and achieve the desired outcome. However, there is a better implementation that has many benefits and does stand out as the better choice, 

```ruby
def each
    return to_enum(:each) unless block_given?

    counter = 0
    while counter < @list.length
      yield(@list[counter])
      counter += 1
    end
    @list
  end
```


and here's why:

Flexibility: The suggested implementation provides more flexibility by allowing the method to be used both with and without a block. When no block is given, it returns an enumerator using to_enum(:each). This is a powerful feature, as it enables users of your class to utilize the method in various scenarios and with different needs.

Readability: The suggested implementation is more explicit in terms of its iteration process. By using a while loop and a counter variable, it's easier to understand that the method is iterating through the @list array and yielding each element in succession. This improves the overall readability and maintainability of your code.

Performance: The suggested implementation could potentially be more efficient in terms of performance. Using the while loop directly on the @list array is likely to have better performance than relying on the @list.each(&block) method, which might involve some additional method call overhead.

Consistency: By handling the iteration explicitly, the suggested implementation maintains consistency within the method itself. Your implementation relies on @list.each(&block) to perform the iteration, which might feel less cohesive, especially when the same behavior can be achieved more directly within the method.

In conclusion, the suggested implementation is the recommended approach due to its flexibility, readability, potential performance benefits, and overall code consistency. It showcases your understanding of different approaches and your ability to choose the most suitable one for the task. Keep up the good work! If you have any questions or need further feedback, feel free to reach out. Happy coding!

**Review by bobb-Rob — APPROVED** [review]
## STATUS: APPROVED ✔️ 🍾 🍾 🍾 🥳 🥳 🥳 

Hi @danielochuba , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights
✅ All linters checks are ok
✅ The correct workflow is used. 
✅ Very descriptive Pull Request(PR)
✅ Your work is well documented.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
